go/consensus/tendermint/apps/beacon/backend_scrape.go

```diff
@@ -0,0 +1,502 @@
+			Epoch: epoch,
+		}
+		scrapeState = newState
+	} else {
+		// The previous attempt to generate a beacon for this epoch failed.
+		scrapeState.Round++
+	}
+
+	// Draw participants.
+	schedulerState := schedulerState.NewMutableState(ctx.State())
+	registryState := registryState.NewMutableState(ctx.State())
+	validators, err := schedulerState.CurrentValidators(ctx)
+	if err != nil {
+		return fmt.Errorf("beacon: failed to get current validators: %w", err)
+	}
+	candidateParticipants, err := validatorsByVotingPower(validators)
+	if err != nil {
+		return fmt.Errorf("beacon: failed to sort current validators: %w", err)
+	}
+
+	numParticipants := int(params.Parameters.SCRAPEParameters.Participants)
+	participants := make([]scrape.Point, 0, numParticipants)
+	participantIDs := make([]signature.PublicKey, 0, numParticipants)
+	for _, validatorID := range candidateParticipants {
+		var node *node.Node
+		node, err = registryState.NodeByConsensusAddress(ctx, validatorID[:])
+		if err != nil || node.Beacon == nil {
+			continue
+		}
+		if scrapeState.BadParticipants[node.ID] {
+			continue
+		}
+
+		participants = append(participants, node.Beacon.Point)
+		participantIDs = append(participantIDs, node.ID)
+		if len(participants) == numParticipants {
+			break
+		}
+	}
+	if l := len(participants); l < numParticipants {
+		return fmt.Errorf("beacon: insufficient beacon participants: %d (want %d)", l, numParticipants)
+	}
+
+	// Initialize the SCRAPE state.
+	if scrapeState.Instance, err = scrape.New(&scrape.Config{
+		Participants:  participants,
+		Threshold:     int(params.Parameters.SCRAPEParameters.Threshold),
+		PVSSThreshold: int(params.Parameters.SCRAPEParameters.PVSSThreshold),
+	}); err != nil {
+		return fmt.Errorf("beacon: failed to initialize SCRAPE instance: %w", err)
+	}
+	scrapeState.Participants = participantIDs
+
+	// Derive the deadlines.
+	height := ctx.BlockHeight() + 1 // Current height is ctx.BlockHeight() + 1
+	scrapeState.CommitDeadline = height + params.Parameters.SCRAPEParameters.CommitInterval
+	scrapeState.RevealDeadline = height + params.Parameters.SCRAPEParameters.RevealInterval
+	scrapeState.TransitionHeight = scrapeState.RevealDeadline + params.Parameters.SCRAPEParameters.TransitionDelay
+
+	scrapeState.Height = height
+	scrapeState.State = beacon.StateCommit
+	if err := state.SetSCRAPEState(ctx, scrapeState); err != nil {
+		return fmt.Errorf("beacon: failed to set SCRAPE state: %w", err)
+	}
+
+	impl.doEmitSCRAPEEvent(ctx, scrapeState)
+
+	ctx.Logger().Info("initializing SCRAPE round",
+		"epoch", epoch,
+		"round", scrapeState.Round,
+	)
+
+	return nil
+}
+
+func (impl *backendSCRAPE) ExecuteTx( //nolint: gocyclo
+	ctx *api.Context,
+	state *beaconState.MutableState,
+	params *beaconState.ConsensusParameters,
+	tx *transaction.Transaction,
+) error {
+	if tx.Method != beacon.MethodSCRAPEReveal && tx.Method != beacon.MethodSCRAPECommit {
+		return fmt.Errorf("beacon: invalid method: %s", tx.Method)
+	}
+
```

Comment:
I think you could make sure that each validator only proposes its own commits/reveals by rejecting transactions in case `ctx.IsCheckOnly()` is true and the `staking.NewAddress(ctx.TxSigner())` is not equal to `ctx.AppState().OwnTxSignerAddress()`. This will prevent other validator's commits/reveals from being included in the local mempool and thus in the blocks we propose.

Then the only other required thing would be to prevent the mempool from gossiping commit/reveal txes to other peers (as other proposers could otherwise still include them). (We don't need to do this second part in this PR to keep it simpler.)

Reply:
Done.

Reply:
It's a problem if there is a large number of commits and or reveals in a single block, because commit/reveal processing is rather CPU intensive.